### ISSUE TITLE
Unwanted bouncing animation on iOS when content of ListView shrinks

### ISSUE THREAD

**Issue by mehmetf** [flutter/flutter#55610]
Internal: b/154880618

This is only reproduced in iOS, probably because of the bouncing scrolling physics. We do need the bouncing effect when user scrolls, but it should be disabled when a list item changes size (I don't think native iOS exhibits this behavior).

Min reproduction case:

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(
    MaterialApp(
      home: Scaffold(
        appBar: AppBar(title: Text('ExpansionTile')),
        body: ListView.builder(
          itemBuilder: (BuildContext context, int index) => index == 0
              ? ExpandingBox()
              : Container(height: 300, color: Colors.red),
          itemCount: 2,
        ),
      ),
    ),
  );
}

class ExpandingBox extends StatefulWidget {
  @override
  _ExpandingBoxState createState() => _ExpandingBoxState();
}

class _ExpandingBoxState extends State<ExpandingBox> {
  var _height = 400.0;

  @override
  Widget build(BuildContext context) {
    return Container(
      height: _height,
      color: Colors.green,
      child: Align(
        alignment: Alignment.bottomCenter,
        child: FlatButton(
          child: Text('Collapse'),
          onPressed: () =>
              setState(() => _height = _height == 400 ? 1200 : 400),
        ),
      ),
    );
  }
}
```

reproduction steps:
1) launch the code
1) press the "Collapse" button once to expand the ListView
1) press the "Collapse" button a second time to shrink the ListView
1) notice the contraction resulted in the undesired animation

See internal bug for a video.

**Comment by mehmetf** [comment]
Critical as the app is failing UX review which blocks launch.

**Comment by xster** [comment]
@khorshijie do you have a mock for the desired visual effect? Should it just instantly snap?

**Comment by khorshijie** [comment]
I don't have a mock. 

I wrote a similar minimal reproduction case on native iOS, and it seems like the view just instantly snaps (attached video in drive link). I think we can mimic the native behavior here.

https://drive.google.com/file/d/1hBQIBuRTFdsrlsYYGCxuLAEzV0fg04EE/view?usp=sharing

**Comment by gaaclarke** [comment]
This doesn't reproduce on `1.18.0-9.0.pre.46` on iOS 13.3

![weird_bounce_sim](https://user-images.githubusercontent.com/30870216/80624869-2646dc80-8a01-11ea-8733-86a8c87da624.gif)

Do you have any more information about where it appears?  Flutter version, iOS version, different reproduction code?

**Comment by gaaclarke** [comment]
I was able to reproduce it.  Some key information was missing, the undesired animation happens the second time you press the button, you have to press it twice.  Thus the animation is happening as a result of the contraction of the listview height.

**Comment by gaaclarke** [comment]
This appears to be what is driving the animation:
```
flutter: animateWith BouncingScrollSimulation(leadingExtent: 0.0, trailingExtent: 0.0) 
#0      AnimationController.animateWith (package:flutter/src/animation/animation_controller.dart:694:49)
#1      new BallisticScrollActivity (package:flutter/src/widgets/scroll_activity.dart:514:9)
#2      ScrollPositionWithSingleContext.goBallistic (package:flutter/src/widgets/scroll_position_with_single_context.dart:148:21)
#3      IdleScrollActivity.applyNewDimensions (package:flutter/src/widgets/scroll_activity.dart:149:14)
#4      ScrollPosition.applyNewDimensions (package:flutter/src/widgets/scroll_position.dart:526:14)
#5      ScrollPositionWithSingleContext.applyNewDimensions (package:flutter/src/widgets/scroll_position_with_single_context.dart:106:11)
#6      ScrollPosition.applyContentDimensions (package:flutter/src/widgets/scroll_position.dart:496:7)
#7      RenderViewport.performLayout (package:flutter/src/rendering/viewport.dart:1357:20)
#8      RenderObject._layoutWithoutResize (package:flutter/<…>
```

**Comment by xster** [comment]
I suppose semantically, to have the strict iOS behavior, it's IdleScrollActivity.applyNewDimensions that shouldn't create any ballistic simulations. But it's a big scope work anticipating all the potential consequences of it.

**Comment by gaaclarke** [comment]
I've seen the same animation on Android, it doesn't appear to be iOS only.  Is it that this animation is expected on Android and not on iOS?

**Comment by xster** [comment]
I haven't tested it but you might right that it could happen on Android too. https://github.com/flutter/flutter/blob/46a5c550dbfad8273d33871bbcf46c5eb4a5afa6/packages/flutter/lib/src/widgets/scroll_simulation.dart#L150. @mehmetf you said it didn't repro on Android? 

I think the solution is still semantically the same (on IdleScrollActivity) but still slightly risky.

**Comment by mehmetf** [comment]
I believe this applies to Android as well but perhaps it is less jarring (glow vs bounce).

**Comment by gaaclarke** [comment]
Here is the lay of the land.  I had to draw it up because it's a lot to keep in your head:
![bounce_sequence](https://user-images.githubusercontent.com/30870216/80649215-73897500-8a26-11ea-8e33-7fc03fec6fff.png)
![bounc_class](https://user-images.githubusercontent.com/30870216/80649230-7c7a4680-8a26-11ea-94fb-8d892e476f31.png)

It seems like we can create a new method on ScrollActivityDelegate that is `applyNewDimensions` that will pop the ScrollPositionWithSingleContext's scroll position to the extent, instead of doing that through an animation.

**Comment by gaaclarke** [comment]
It appears that when `ScrollPositionWithSingleContext.goBalistic()` is called `ScrollPositionWithSingleContext.maxScrollExtent == 0`.  This probably contributes to the oddness of the animation.  I speculate the end result of a contraction that causes the offset to be out-of-bounds will always be `offset == 0.0`.  It just so happens that that is the correct end state for the sample, but wouldn't be for a contraction that happens lower in a longer ListView.

**Comment by gaaclarke** [comment]
I tried to use a `DrivenScrollActivity` to jump directly to offset zero instead of using a `BallisticScrollActivity`.  It gets the results we want but it introduces a single frame blip sometimes.

**Comment by gaaclarke** [comment]
Using a `BalisticScrollActivity` with a `Simulation` that just completes and goes to one offset also sometimes results in a singular frame being white, the aforementioned "blip".

**Comment by gaaclarke** [comment]
https://github.com/flutter/flutter/pull/55993 does the jump we want, but here is the visual glitch that happens sometimes.  I'll shop the draft around to see if someone has some idea on how to get rid of it:
![ezgif-4-41fd3e416c37](https://user-images.githubusercontent.com/30870216/80655699-2f04d600-8a34-11ea-98db-2fe2a2da118d.gif)

**Comment by gaaclarke** [comment]
Even just directly calling `jumpTo(0.0)` as a result of `applyNewDimensions` results in a half second of white being rendered before the jump.  I'm now skeptical we can eliminate this animation, the library really wants an animation here.  It is starting to look like the amount of plumbing to remove may be massive.

**Comment by Hixie** [comment]
What's going on is that when the button is pressed the second time, the scroll view suddenly finds itself with a massively out-of-bounds position, and it needs to do something about it. I'm surprised to hear that it only reproduces on iOS; I can get it to happen on iOS and on Android, it's not platform-specific code as far as I recall.

When the cause is a much smaller delta, e.g. because you're looking at a list of items and the last item was removed, then the animation is clearly desireable. I would argue that the problem here is that the first item in the list didn't animate when it shrank (though I guess to look good it would have to animate with exactly the same logic as the ballistic scroll to get us back to within bounds, which would be... tricky to pull off).

Forcing the list to never overscroll should be possible by providing a ScrollPhysics that implements a very strict applyBoundaryConditions.

**Comment by Hixie** [comment]
Hm, actually, I guess we never ask the physics to do the scroll in this case... that's probably a mistake. It means that applyBoundaryConditions doesn't get a chance to enforce the rule...

**Comment by Hixie** [comment]
Thinking out loud, it's sort of weird that we call ScrollActivity.applyNewDimensions and rely on it calling goBallistic...

**Comment by gaaclarke** [comment]
> Hm, actually, I guess we never ask the physics to do the scroll in this case... that's probably a mistake. It means that applyBoundaryConditions doesn't get a chance to enforce the rule...

That matches what I'm seeing in my PR, I print out the `overscroll` in `setPixels` and it always prints 0 even though sometimes a white frame is rendered.

**Comment by Hixie** [comment]
Depending on what we actually want to have happen here, it's arguably the case that we should be applying a correction to the offset during layout. Anything we do after layout will be too late and will show the white frame regardless of how clever we are, since it happens after the offending layout.

That suggests that we should have the viewport contact the physics somehow (well, the activity, actually, since we still want to allow iOS-style bounce, and the activity should contact the physics) at the end of layout to get final approval...

**Comment by Hixie** [comment]
aha, that exists already, it's called `offset.applyContentDimensions`

**Comment by Hixie** [comment]
and that already calls applyNewDimensions... we're close to the code path we need here, just needs a bit of help.

**Comment by gaaclarke** [comment]
@Hixie Are you talking about this?
```diff
--- a/packages/flutter/lib/src/widgets/scroll_position.dart
+++ b/packages/flutter/lib/src/widgets/scroll_position.dart
@@ -479,30 +479,37 @@ abstract class ScrollPosition extends ViewportOffset with ScrollMetrics {
     context.setSemanticsActions(_semanticActions);
   }
 
   @override
   bool applyContentDimensions(double minScrollExtent, double maxScrollExtent) {
     assert(minScrollExtent != null);
     assert(maxScrollExtent != null);
     if (!nearEqual(_minScrollExtent, minScrollExtent, Tolerance.defaultTolerance.distance) ||
         !nearEqual(_maxScrollExtent, maxScrollExtent, Tolerance.defaultTolerance.distance) ||
         _didChangeViewportDimensionOrReceiveCorrection) {
       assert(minScrollExtent != null);
       assert(maxScrollExtent != null);
       assert(minScrollExtent <= maxScrollExtent);
       _minScrollExtent = minScrollExtent;
       _maxScrollExtent = maxScrollExtent;
+      if (pixels > _maxScrollExtent) {
+        correctPixels(_maxScrollExtent);
+      }
+      if (pixels < _minScrollExtent) {
+        correctPixels(_minScrollExtent);
+      }
       _haveDimensions = true;
       applyNewDimensions();
       _didChangeViewportDimensionOrReceiveCorrection = false;
```

That still renders off the bounds of the listview for a half second... maybe it's fighting with applyNewDimensions... If this is what you were thinking I can poke around more tomorrow.

**Comment by Hixie** [comment]
The key is returning false from that method, to retrigger layout. But only doing it during idle activities. And having the physics actually decide what to do.

The tricky part is that the activities don't have access to the physics, which is weird. I'm trying to figure out if we should let them have access or not.

**Comment by Hixie** [comment]
Observation: if this happens while you're overscrolling, you really want the position to be snapped not to zero, but to the same delta from the max extent (or min extent when underscrolling) as it had before. We don't seem to track the old value, sadly...

**Comment by Hixie** [comment]
https://github.com/flutter/flutter/compare/master...Hixie:scroll_snap is what I came up with. It solves the problem in the case where the user isn't scrolling and everything is idle, but in the case where the user is scrolling, or there's a ballistic or animated scroll happening, it still has the same problem as before.

(Also, it needs tests. I did update the docs though.)

**Comment by gaaclarke** [comment]
@Hixie Thanks, a few questions on the PR:
1) why even delegate to ScrollPhysics if the logic is unconditionally the same?
2) in the case that we correctPixels, shouldn't we go to an IdleActivity?
3) do you want me to pick it up and bring it to completion?

@Hixie @xster Outstanding questions for issue:
1) Given Hixie's approach are we confident making the change with respect to Risk vs Reward?
2) The animation appears on Android and iOS, the issue only complains about iOS, do we want to remove the animation for both platforms?
3) Replacing the incorrect animation with an animation of length zero is less invasive, but sometimes renders one frame wrong.  Is that a reasonable compromise? What is more wrong, a 1 second animation or a 1 frame animation sometimes?

**Comment by xster** [comment]
What I was saying on internal chat yesterday is that since we have implemented a feature that no one asked for (animate to new extent edge on applyNewDimensions when pixel is outside the edge), we can just globally remove that mechanism by just letting applyNewDimensions call correctPixels itself when pixel is outside min/maxScrollExtent (and let it return a bool). 

But Ian's solution is better since it kept it customizable per physics from the user. 

> The animation appears on Android and iOS, the issue only complains about iOS, do we want to remove the animation for both platforms?

Yes, I think clamping it for both physics is sensible.

> Replacing the incorrect animation with an animation of length zero is less invasive, but sometimes renders one frame wrong. Is that a reasonable compromise? What is more wrong, a 1 second animation or a 1 frame animation sometimes?

We should make sure there isn't a wrong frame.

**Comment by gaaclarke** [comment]
> We should make sure there isn't a wrong frame.

It sounded like you and hixie were reluctant to accept the collateral implications of such a change, that's why I ask.  I know what is obviously better.  I don't have enough experience with this code to know how risky of a change this is.

> But Ian's solution is better since it kept it customizable per physics from the user.

Ok, I'll preserve delegating to the ScrollPhysics.

> Yes, I think clamping it for both [Android and iOS] physics is sensible.

Roger.

**Comment by Hixie** [comment]
> @Hixie Thanks, a few questions on the PR:
> 
> 1. why even delegate to ScrollPhysics if the logic is unconditionally the same?

It's not the same. On iOS, for example, if you end up in this situation while overscrolling, the behaviour would be different than on Android. (Also, I suspect people might wont to override it to be different.) The branch I posted doesn't provide a complete solution.

> 2. in the case that we correctPixels, shouldn't we go to an IdleActivity?

That would mean that if anything changed size while you were scrolling, your scroll would be canceled.


> 3. do you want me to pick it up and bring it to completion?

Sure!


> 2. The animation appears on Android and iOS, the issue only complains about iOS, do we want to remove the animation for both platforms?

I'm more confident that we want to remove this for Android than for iOS. On Android it's 100% wrong as far as I can tell. On iOS, the situation is less clear (see my comments above about needing to maintain the delta from min-or-max to position during under/over-scroll, rather than just removing the gap entirely).


> 3. Replacing the incorrect animation with an animation of length zero is less invasive, but sometimes renders one frame wrong.  Is that a reasonable compromise? What is more wrong, a 1 second animation or a 1 frame animation sometimes?

The correct approach is definitely to correct the pixels during layout rather than one frame later. One frame later looks really ugly, arguably worse than having the animation drag it back, IMHO.

**Comment by gaaclarke** [comment]
> That would mean that if anything changed size while you were scrolling, your scroll would be canceled.

That is the current behavior, the resetting animation will cancel any current activity.  Don't we want to preserve that?

With respect to customization, your patch doesn't allow someone to get the behavior we are removing for example, should we allow someone that level of customization without subclassing ScrollView?

**Comment by gaaclarke** [comment]
Note: I'm running into an interesting impasse.  The pull-down-to-refresh mechanism is being broken by removing this animation.  When you pull down and an item is inserted you find that your content size has changed and you are at a negative offset out-of-bounds so it jumps back instead of animating like it should.  The way this mechanism is setup is with something like a `CupertinoSliverRefreshControl` with a standard `CustomScrollView` and `ScrollPhysics` so there isn't an easy way to replace this behavior.

We may not be able to get rid of the animation as the default behavior but we can make it possible to have with a custom `ScrollPhysics`.

**Comment by Hixie** [comment]
Maybe with the bouncy physics we can only clamp when it's past the max extent?

**Comment by xster** [comment]
That would work but it's a very specific behavior to explain in the API docs :)

**Comment by gaaclarke** [comment]
> Maybe with the bouncy physics we can only clamp when it's past the max extent?

The thought crossed my mind then I arrived where @xster is.  It's just too oddball to treat them differently.

There was also some weirdness related to ScrollViews who don't know their extents at construction time also.  Their max extents were bouncing between infinity and the calculated max which was also tripping things up.  I'm going to pull back and make it so you can turn the animation is off but not touch the default behavior.  Maybe someone more adept can solve this Hellraiser puzzle box (apparently it's called a 'Lament Configuration').

**Comment by Hixie** [comment]
I think what would be most helpful if you want to hand this off to someone else would be to create test cases for all the cases you ran into.

**Comment by Hixie** [comment]
(I don't think we should land a PR that just turns off the animation.)

**Comment by gaaclarke** [comment]
@Hixie 

> I think what would be most helpful if you want to hand this off to someone else would be to create test cases for all the cases you ran into.

I've just put a PR in review (https://github.com/flutter/flutter/pull/55993).  It adds tests that cover what happens when the extents shrink, that behavior wasn't asserted elsewhere in tests.

The things I ran into were existing test cases.  The whole approach of removing this animation is antithetical to a few different variants of ScrollViews.  I've listed them above.  Unfortunately the way ScrollView's are constructed there doesn't appear to be a way to not break those and make the example code behave the way we want verbatim.  Maybe someone else could find a way but after days of research I couldn't find a way.

> (I don't think we should land a PR that just turns off the animation.)

No one is suggesting that.  The PR I put up for review allows users to turn off the animation with a custom ScrollPhysics without affecting current default behavior.
